[FEATURE][needs-docs] import/export section of vector layer style

3nids opened the pull request:
implements https://github.com/qgis/QGIS-Enhancement-Proposals/issues/125

This moves some things to `QgsMapLayer::read/writeCommonStyle` to avoid code duplication in `read/writeLayerXML` and `import/exportNamedStyle`.
Flags have been added to `QgsMapLayer` to represent style categories of what is export in styles.

3nids committed:
add API to select section of layer style when reading/writing

m-kuhn commented on src/core/qgsmaplayer.h:
```diff
@@ -736,7 +737,8 @@ class CORE_EXPORT QgsMapLayer : public QObject
      */
-    virtual void exportNamedStyle( QDomDocument &doc, QString &errorMsg SIP_OUT ) const;
+    virtual void exportNamedStyle( QDomDocument &doc, QString &errorMsg SIP_OUT, QgsReadWriteContext &context,
+                                   QgsMapLayerStyle::StyleCategories categories = QgsMapLayerStyle::All ) const;
```
Technically speaking, this is an API break (if someone subclassed and overwrote).
I guess a mapping could be done but it's definitely not trivial.

Question: is this even actively used? Then at least there would be an example on which to base this.

@nyalldawson @wonder-sk (others?)

nyalldawson replied:
I don't think overriding layer methods is really intended use of the API. If we could, I'd like them to be marked "final" in the vector/raster classes. But sip doesn't like this, so we're forced with this annoying situation.

Anyway, I'm ok with breaking non-intended use of API.

nyalldawson replied:
If it works, I'm more than happy to use it!

m-kuhn commented:
Looks good to me so far :+1:

nyalldawson commented on src/app/qgisapp.cpp:
```diff
@@ -9730,7 +9731,8 @@ void QgisApp::duplicateLayers( const QList<QgsMapLayer *> &lyrList )
     QDomDocument style;
-    selectedLyr->exportNamedStyle( style, errMsg );
+    QgsReadWriteContext context;
+    selectedLyr->exportNamedStyle( style, errMsg, context );
```
Why are these changes needed?

3nids replied:
exportNamedStyle was currently using a context created in QgsMapLayer. I think it should be created from the app to be able to report messages back.
Even if not used at the moment, this makes the code as the methods were initially thought.

nyalldawson commented on src/core/qgsmaplayer.cpp:
```diff
@@ -306,20 +302,6 @@ bool QgsMapLayer::readLayerXml( const QDomElement &layerElement,  QgsReadWriteCo
     }
   }
 
-  // use scale dependent visibility flag
```
These changes - while desirable - make me nervous. There's potentially signals getting thrown in the moved code and the change of property restoration order here may have unintentional tricky side effects. I don't think that should block this refactoring... But proceed with extreme caution!

3nids committed:
implement filtering of categories in reading/writing

3nids committed:
use qgsEnumMap

3nids committed:
select categories in copy style in legend menu entry

not working yet, just set the menu entries
QgsMapLayerStyle::StyleCategory has moved to QgsMapLayer to avoid making QgsMapLayerStyle a QObject and they are mostly used in QgsMapLayer

3nids committed:
add icons and export only selected category

3nids commented:
![image](https://user-images.githubusercontent.com/127259/45444301-20b64a80-b695-11e8-9909-5d4a0d76d6c3.png)

3nids committed:
fix sip include

3nids committed:
categories in paste style

saberraz commented:
Nice feature!

IMO it will be better to have dedicated buttons on toolbar for copy and pasting style. It is very nested and not very easy to use in this form.

How about having a pint-brush type icon (similar to Word processing and spreadsheets):
https://support.content.office.net/en-us/media/12598786-773c-4e92-be5f-88d766f115ec.jpg
with drop-down options.

3nids commented:
@saberraz thanks. There is more details in the QEP. There will be a dialog for advanced usage. I don't think this feature is worth its dedicated icon (I will also implement entries in the layer menu).

nyalldawson commented on src/app/qgsapplayertreeviewmenuprovider.cpp:
```diff
@@ -316,11 +316,51 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
+          for ( QgsMapLayer::StyleCategory category : categories )
+          {
+            QgsMapLayer::ReadableStyleCategory readableCategory = QgsMapLayer::readableStyleCategory( category );
+            QAction *copyAction = new QAction( readableCategory.icon(), readableCategory.name() );
```
this is leaked

3nids replied:
fixed

nyalldawson commented on src/app/qgsapplayertreeviewmenuprovider.cpp:
```diff
@@ -316,11 +316,51 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
+        }
+        else
+        {
+          menuStyleManager->addAction( tr( "Copy Style" ), app, SLOT( copyStyle() ) );
```
would be better to split this call to avoid the old style connection required by the shortcut addAction method

nyalldawson commented on src/core/qgsmaplayer.cpp:
```diff
@@ -67,6 +67,51 @@ QString QgsMapLayer::extensionPropertyType( QgsMapLayer::PropertyType type )
+    case LayerConfiguration:
+      return ReadableStyleCategory( tr( "Layer Configuration" ),
+                                    QgsApplication::getThemeIcon( QStringLiteral( "/propertyicons/system.svg" ) ),
+                                    tr( "Flags, display expression, read-only" ) );
```
"Flags" will mean nothing to users. Better to be explicit here.

nyalldawson commented on src/core/qgsmaplayer.cpp:
```diff
@@ -67,6 +67,51 @@ QString QgsMapLayer::extensionPropertyType( QgsMapLayer::PropertyType type )
+      return ReadableStyleCategory( tr( "Diagrams" ),
+                                    QgsApplication::getThemeIcon( QStringLiteral( "/propertyicons/diagram.svg" ) ) );
+    case AttributeTable :
+      return ReadableStyleCategory( tr( "Attribute Table" ),
```
"Attribute Table Settings"?

nyalldawson commented on src/core/qgsmaplayer.cpp:
```diff
@@ -67,6 +67,51 @@ QString QgsMapLayer::extensionPropertyType( QgsMapLayer::PropertyType type )
+    case Symbology :
+      return ReadableStyleCategory( tr( "Symbology" ),
+                                    QgsApplication::getThemeIcon( QStringLiteral( "/propertyicons/symbology.svg" ) ) );
+    case Labels :
```
Should 3D be it's own category? (I think yes)

nyalldawson commented on src/core/qgsmaplayer.h:
```diff
@@ -133,6 +135,52 @@ class CORE_EXPORT QgsMapLayer : public QObject
+    enum StyleCategory
+    {
+      LayerConfiguration = 1 << 0, //!< Flags, display expression, read-only
+      Symbology          = 1 << 1,
```
Can you add dox for all these? I realise some are self explanatory, but it's the precedent

nyalldawson commented on src/core/qgsmaplayer.h:
```diff
@@ -133,6 +135,52 @@ class CORE_EXPORT QgsMapLayer : public QObject
+    struct ReadableStyleCategory
+    {
+      public:
+        ReadableStyleCategory( const QString &name, const QString &toolTip = QString() )
```
Constructor dox, getter dox missing

nyalldawson commented on src/core/qgsmaplayer.cpp:
```diff
@@ -67,6 +67,51 @@ QString QgsMapLayer::extensionPropertyType( QgsMapLayer::PropertyType type )
+      return ReadableStyleCategory( tr( "Custom Properties" ),
+                                    QgsApplication::getThemeIcon( QStringLiteral( "/mActionOptions.svg" ) ) );
+    case AllCategories :
+      return ReadableStyleCategory( tr( "All categories" ) );
```
All Categories (capitalization)

3nids committed:
better API doc, own category for 3D, fix leak

3nids commented:
some open questions:
* why is QgsMapLayer::exportNamedStyle virtual? does it make any sense? shall I make it non virtual?
* the are read/writeCustomSymbology signals in vector layer:
  * should I embed the emit signals in a test for a category? 
  * if yes, symbology? new category? custom properties???

3nids committed:
fix old style or useless casts

3nids committed:
add test to control that AllStyleCategories is actually complete

3nids committed:
create dialog for saving different styles

3nids committed:
single dialog to save all types (QML, SLD, DataBase)

3nids committed:
fix documentation

3nids committed:
single dialog to load style (QML, SLD, DB)

nyalldawson commented on src/app/qgsvectorlayersavestyledialog.h:
```diff
@@ -0,0 +1,62 @@
+    void readUiFileContent( const QString &filePath );
+
+  private:
+    QgsVectorLayer *mLayer;
```
initialize to nullptr

nyalldawson commented on src/app/qgsvectorlayersavestyledialog.cpp:
```diff
@@ -0,0 +1,196 @@
+    mStyleCategoriesListWidget->setEnabled( type == QgsVectorLayerProperties::QML );
+    mFileWidget->setFilter( type == QgsVectorLayerProperties::QML ? tr( "QGIS Layer Style File (*.qml)" ) : tr( "SLD File (*.sld)" ) );
+  } );
+  mStyleTypeComboBox->addItem( tr( "as QGIS QML style file" ), QgsVectorLayerProperties::QML );
```
HIG guidelines is that combobox entries are title case: "As QGIS QML Style File"

nyalldawson commented on src/app/qgsvectorlayersavestyledialog.cpp:
```diff
@@ -0,0 +1,196 @@
+    {
+      if ( mFileWidget->filePath().isEmpty() )
+      {
+        QMessageBox::information( this, tr( "Save Style in file" ), tr( "Select a file destination." ) );
```
"Save Style in File"

nyalldawson replied:
(sorry... just getting in before @DelazJ ;) )

DelazJ replied:
:smiley:

nyalldawson reviewed (commented):
This is super nice to have! Lovely stuff :tada:

nyalldawson commented:
Minor thing looking at that screenshot - I think it'd work better if "Categories" was a group box, instead of the centered label on the left

+---Categories ------------------------+
|  [x] Layer Configuration              |

etc

3nids committed:
fix offline editing

3nids committed:
fixes: HIG guidelines, sipify, initialize, doxygen warnings

fix doxygen warning + docs

sipify

fix doxygen warning

initialize

nirvn commented:
@3nids , great job, this'll be really useful.

3nids committed:
[sipify] fix annotations in multiline virtual methods

signedav commented:
@3nids Nice :slightly_smiling_face:

nyalldawson commented:
Merged because @3nids is :sleeping:

DelazJ commented:
Looks like the merge commit misses the appropriate tags to generate commits in docs (and changelog?). I can create one in docs but if changelog misses it, someone has to do it (still not sure whether changelog issues are automated or not)
